File: php_secrets_saver/src/SecretsSaver.php
<?php

declare(strict_types=1);

namespace SecretsSaver;

use RuntimeException;

interface DatabaseAdapter
{
    public function readEncryptedRow(): ?array;
    public function upsertEncryptedRow(array $content): void;
}

final class InvalidKeyOrCorruptedDataException extends RuntimeException
{
    public function __construct()
    {
        parent::__construct('Invalid key or corrupted data.');
    }
}

final class SecretsSaver
{
    private const DEFAULT_FILENAME = 'secrets.ep';
    private const PBKDF2_ITERS = 600000;
    private const KEY_LEN = 32;
    private const SALT_LEN = 16;
    private const NONCE_LEN = 12;
    private const TAG_LEN = 16;

    private string $filename;
    private ?DatabaseAdapter $dbAdapter;
    private $prompt;

    private ?string $key = null;
    private ?array $data = null;

    private function __construct(string $filename, ?DatabaseAdapter $dbAdapter = null, ?callable $prompt = null)
    {
        $this->filename = $filename === '' ? self::DEFAULT_FILENAME : $filename;
        $this->dbAdapter = $dbAdapter;
        $this->prompt = $prompt ?? static fn(string $location): string => self::defaultPrompt($location);

        if (!$this->exists()) {
            $this->getKey();
            $this->data = [];
            $this->save();
        }
    }

    public static function newFile(string $filename = self::DEFAULT_FILENAME, ?callable $prompt = null): self
    {
        return new self($filename, null, $prompt);
    }

    public static function newDatabase(DatabaseAdapter $adapter, ?callable $prompt = null): self
    {
        return new self(self::DEFAULT_FILENAME, $adapter, $prompt);
    }

    public function setSecret(string $key, string $value): void
    {
        $this->ensureLoaded();
        $this->data[$key] = $value;
        $this->save();
    }

    public function getSecret(string $key): ?string
    {
        $this->ensureLoaded();
        return $this->data[$key] ?? null;
    }

    public function listSecrets(): array
    {
        $this->ensureLoaded();
        $keys = array_keys($this->data);
        sort($keys, SORT_STRING);
        return $keys;
    }

    public function clearDatabase(): void
    {
        $this->data = [];
        $this->save();
    }

    private function exists(): bool
    {
        if ($this->dbAdapter !== null) {
            return $this->dbAdapter->readEncryptedRow() !== null;
        }
        return is_file($this->filename);
    }

    private function getKey(): string
    {
        if ($this->key !== null) {
            return $this->key;
        }

        $location = $this->dbAdapter !== null ? 'database' : $this->filename;
        $prompt = $this->prompt;
        $this->key = (string)$prompt($location);
        return $this->key;
    }

    private function deriveKey(string $password, string $salt): string
    {
        return hash_pbkdf2('sha256', $password, $salt, self::PBKDF2_ITERS, self::KEY_LEN, true);
    }

    private function loadRaw(): array
    {
        if ($this->dbAdapter !== null) {
            $row = $this->dbAdapter->readEncryptedRow();
            if ($row === null) {
                throw new RuntimeException('Secrets not found in database.');
            }
            return $row;
        }

        $raw = file_get_contents($this->filename);
        if ($raw === false) {
            throw new RuntimeException('Unable to read secrets file.');
        }
        $content = json_decode($raw, true);
        if (!is_array($content)) {
            throw new RuntimeException('Invalid payload format.');
        }
        return $content;
    }

    private function saveRaw(array $content): void
    {
        if ($this->dbAdapter !== null) {
            $this->dbAdapter->upsertEncryptedRow($content);
            return;
        }

        $json = json_encode($content, JSON_UNESCAPED_SLASHES);
        if ($json === false) {
            throw new RuntimeException('Unable to encode encrypted payload.');
        }

        if (file_put_contents($this->filename, $json) === false) {
            throw new RuntimeException('Unable to write secrets file.');
        }
    }

    private function load(): void
    {
        $content = $this->loadRaw();

        $salt = base64_decode((string)$content['salt'], true);
        $nonce = base64_decode((string)$content['nonce'], true);
        $ciphertextPlusTag = base64_decode((string)$content['ciphertext'], true);
        if ($salt === false || $nonce === false || $ciphertextPlusTag === false || strlen($ciphertextPlusTag) < self::TAG_LEN) {
            throw new RuntimeException('Invalid payload encoding.');
        }

        $ciphertext = substr($ciphertextPlusTag, 0, -self::TAG_LEN);
        $tag = substr($ciphertextPlusTag, -self::TAG_LEN);

        $key = $this->deriveKey($this->getKey(), $salt);
        $plaintext = openssl_decrypt(
            $ciphertext,
            'aes-256-gcm',
            $key,
            OPENSSL_RAW_DATA,
            $nonce,
            $tag,
            ''
        );

        if ($plaintext === false) {
            $this->key = null;
            throw new InvalidKeyOrCorruptedDataException();
        }

        $parsed = json_decode($plaintext, true);
        if (!is_array($parsed)) {
            throw new RuntimeException('Invalid decrypted payload.');
        }

        $this->data = $parsed;
    }

    private function ensureLoaded(): void
    {
        if ($this->data !== null) {
            return;
        }

        if ($this->exists()) {
            $this->load();
        } else {
            $this->data = [];
        }
    }

    private function save(): void
    {
        $this->ensureLoaded();

        $salt = random_bytes(self::SALT_LEN);
        $nonce = random_bytes(self::NONCE_LEN);
        $key = $this->deriveKey($this->getKey(), $salt);

        $plaintext = json_encode($this->data, JSON_UNESCAPED_SLASHES);
        if ($plaintext === false) {
            throw new RuntimeException('Unable to encode plaintext payload.');
        }

        $tag = '';
        $ciphertext = openssl_encrypt(
            $plaintext,
            'aes-256-gcm',
            $key,
            OPENSSL_RAW_DATA,
            $nonce,
            $tag,
            '',
            self::TAG_LEN
        );

        if ($ciphertext === false) {
            throw new RuntimeException('Encryption failed.');
        }

        $this->saveRaw([
            'salt' => base64_encode($salt),
            'nonce' => base64_encode($nonce),
            'ciphertext' => base64_encode($ciphertext . $tag),
        ]);
    }

    private static function defaultPrompt(string $location): string
    {
        if (function_exists('readline')) {
            $line = readline("Enter key for {$location}: ");
            return $line === false ? '' : trim($line);
        }

        fwrite(STDOUT, "Enter key for {$location}: ");
        $line = fgets(STDIN);
        return $line === false ? '' : trim($line);
    }
}

File: ruby_secrets_saver/lib/secrets_saver.rb
# frozen_string_literal: true

require "base64"
require "json"
require "openssl"
require "securerandom"
require "io/console"

module SecretsSaver
  class InvalidKeyOrCorruptedDataError < StandardError
    def initialize
      super("Invalid key or corrupted data.")
    end
  end

  class SecretsSaver
    DEFAULT_FILENAME = "secrets.ep"
    PBKDF2_ITERS = 600_000
    KEY_LEN = 32
    SALT_LEN = 16
    NONCE_LEN = 12

    def self.new_file(filename = DEFAULT_FILENAME, prompt: nil)
      new(filename, db_adapter: nil, prompt: prompt)
    end

    def self.new_db(db_adapter, prompt: nil)
      new(DEFAULT_FILENAME, db_adapter: db_adapter, prompt: prompt)
    end

    def initialize(filename, db_adapter: nil, prompt: nil)
      @filename = (filename.nil? || filename.empty?) ? DEFAULT_FILENAME : filename
      @db_adapter = db_adapter
      @prompt = prompt || method(:default_prompt)
      @key = nil
      @data = nil

      unless exists?
        get_key
        @data = {}
        save
      end
    end

    def set_secret(key, value)
      ensure_loaded
      @data[key] = value
      save
    end

    def get_secret(key)
      ensure_loaded
      @data[key]
    end

    def list_secrets
      ensure_loaded
      @data.keys.sort
    end

    def clear_database
      @data = {}
      save
    end

    private

    def exists?
      if @db_adapter
        !@db_adapter.read_encrypted_row.nil?
      else
        File.file?(@filename)
      end
    end

    def get_key
      return @key if @key

      location = @db_adapter ? "database" : @filename
      @key = @prompt.call(location).to_s
    end

    def derive_key(password, salt)
      OpenSSL::PKCS5.pbkdf2_hmac(password, salt, PBKDF2_ITERS, KEY_LEN, OpenSSL::Digest::SHA256.new)
    end

    def load_raw
      if @db_adapter
        row = @db_adapter.read_encrypted_row
        raise "Secrets not found in database." if row.nil?

        row
      else
        JSON.parse(File.read(@filename, mode: "r:bom|utf-8"))
      end
    end

    def save_raw(content)
      if @db_adapter
        @db_adapter.upsert_encrypted_row(content)
      else
        File.write(@filename, JSON.generate(content))
      end
    end

    def load
      content = load_raw

      salt = Base64.strict_decode64(content.fetch("salt"))
      nonce = Base64.strict_decode64(content.fetch("nonce"))
      ciphertext_plus_tag = Base64.strict_decode64(content.fetch("ciphertext"))

      raise InvalidKeyOrCorruptedDataError if ciphertext_plus_tag.bytesize < 16

      ciphertext = ciphertext_plus_tag[0...-16]
      tag = ciphertext_plus_tag[-16, 16]

      begin
        cipher = OpenSSL::Cipher.new("aes-256-gcm")
        cipher.decrypt
        cipher.key = derive_key(get_key, salt)
        cipher.iv = nonce
        cipher.auth_tag = tag
        cipher.auth_data = ""

        plaintext = cipher.update(ciphertext) + cipher.final
        @data = JSON.parse(plaintext)
      rescue OpenSSL::Cipher::CipherError
        @key = nil
        raise InvalidKeyOrCorruptedDataError
      end
    end

    def ensure_loaded
      return if @data

      if exists?
        load
      else
        @data = {}
      end
    end

    def save
      ensure_loaded

      salt = SecureRandom.random_bytes(SALT_LEN)
      nonce = SecureRandom.random_bytes(NONCE_LEN)

      cipher = OpenSSL::Cipher.new("aes-256-gcm")
      cipher.encrypt
      cipher.key = derive_key(get_key, salt)
      cipher.iv = nonce
      cipher.auth_data = ""

      plaintext = JSON.generate(@data)
      ciphertext = cipher.update(plaintext) + cipher.final
      tag = cipher.auth_tag

      save_raw(
        {
          "salt" => Base64.strict_encode64(salt),
          "nonce" => Base64.strict_encode64(nonce),
          "ciphertext" => Base64.strict_encode64(ciphertext + tag)
        }
      )
    end

    def default_prompt(location)
      print "Enter key for #{location}: "
      if $stdin.tty?
        $stdin.noecho(&:gets).to_s.strip.tap { puts }
      else
        $stdin.gets.to_s.strip
      end
    end
  end
end
